blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc

```diff
@@ -0,0 +1,647 @@
+</div>
+++++
+
+The transaction id is part of the `source` structure of vegetable change events,
+and it is the message key of metadata events.
+Once we've found the corresponding transaction event for a given data change event,
+the `client_date`, `usecase` and `user_name` attributes from the former can be added to the latter:
+
+++++
+<div class="imageblock centered-image">
+    <img src="/images/auditing_output_message.png" style="max-width:100%; margin-bottom:10px; margin-top:10px;" class="responsive-image" alt="Enriched Vegetable Message">
+</div>
+++++
+
+This kind of message transformation is a perfect use case for https://kafka.apache.org/documentation/streams/[Kafka Streams],
+a Java API for implementing stream processing applications on top of on Kafka topics,
+providing operators that let you filter, transform, aggregate and join Kafka messages.
+
+As foundation for our stream processing application we're going to use https://quarkus.io/[Quarkus],
+which is "a Kubernetes Native Java stack tailored for GraalVM & OpenJDK HotSpot, crafted from the best of breed Java libraries and standards".
+
+.Building Kafka Streams with Quarkus
+[NOTE]
+===============================
+Amongst many others, Quarkus comes with an https://quarkus.io/guides/kafka-streams-guide[extension for Kafka Streams],
+which allows to build stream processing applications running on the JVM and as native code compiled ahead-of-time.
+It takes care of the lifecycle of the streaming topology,
+so you don't have to deal with details like registering JVM shutdown hooks,
+awaiting the creation of all input topics and more.
+
+The extension also comes with "live development" support,
+which automatically reloads the stream processing application while you're working on it,
+allowing for very fast turnaround cycles during development.
+===============================
+
+=== The Joining Logic
+
+When thinking about the actual implementation of the enrichment logic,
+a https://kafka.apache.org/23/documentation/streams/developer-guide/dsl-api.html#kstream-kstream-join[stream-to-stream] join might appear as a suitable solution.
+By creating ``KStream``s for the two topics, we may try and implement the joining functionality.
+This creates a problem though in regards to ordering guarantees of the change events.
+By default, Debezium will use a table's primary key as the message key for the corresponding Kafka messages.
+This means that all messages for the same vegetable record will have the same key and thus will go into the same partition of the vegetables Kafka topic.
+This in turn guarantees that a consumer of these events sees all the messages pertaining to the same vegetable record in the exact same order as they were created.
+
+Now, in order to join the two streams, the message key must be the same on both sides.
```

Comment:
You might add just 1 sentence that a stream-stream join would also need a window for a length which is not easy to define in general.

Reply:
Done.